### DISCUSSION TITLE
multi-thread sync support

### DISCUSSION THREAD

**Discussion by jcyuan** [gfx-rs/wgpu#5712] [Q&A] [answered]
Hi guys, i'm planning to develop an app which has a few threads: UI / video player / rendering.
and i'm investigating if wgpu meets my needs to be the main rendering backend, but i'm not sure if it supports this situation:
once the main image is ready (rendered in the rendering thread by wgpu), i need to submit it to the UI thread (UI thread has an Opengl simple rendering logic to show an external image (with VK_external_object extension) as the preview source, but before this, i need to do image sync between these 2 threads.
why i'm asking about this question is because i know webgpu does not support multi-thread iirc?

**Comment by kpreid** [comment, marked as answer]
As long as you're not actually running your application in a web page, `wgpu` supports use from multiple threads. The caveat is that there is currently only one `Queue`, so work cannot actually be submitted to the GPU truly in parallel; there will be some unnecessary synchronization. The internals of `wgpu` also currently have more exclusive locking than is ideal, if I understand correctly. None of this prevents you from using `wgpu` from multiple threads; it just will not be as high-performance as it theoretically could be. And, of course, the GPU itself is still doing its actual work asynchronously.

**Reply by jcyuan** [reply]
Hi thanks, it does make sense, yes for my case one queue is enough because i don't need some complicated multi-queue rendering, and for the final image sync between backends i believe it's just fine as you said.
will try to integrate with wgpu now.   :)
